From 9510454fd66fabbf1b1c4712367c371968afbf36 Mon Sep 17 00:00:00 2001
From: Piyush Mehta <piyush.mehta@xilinx.com>
Date: Tue, 24 Mar 2020 11:45:40 +0530
Subject: [PATCH 1292/1851] usb: dwc3: when enters in d3 state

commit aec0c421cfb5b9598d8b9c58d10e0e522531afef from
https://github.com/Xilinx/linux-xlnx.git

This patch adds wakeup_capable and is_d3 flag handling in suspend function,
when host enters in d3 state and disable clock.

Signed-off-by: Piyush Mehta <piyush.mehta@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/usb/dwc3/dwc3-of-simple.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/dwc3/dwc3-of-simple.c b/drivers/usb/dwc3/dwc3-of-simple.c
index 50d7e20a2d39..49f7edcff395 100644
--- a/drivers/usb/dwc3/dwc3-of-simple.c
+++ b/drivers/usb/dwc3/dwc3-of-simple.c
@@ -647,8 +647,13 @@ static int __maybe_unused dwc3_of_simple_suspend(struct device *dev)
 {
 	struct dwc3_of_simple *simple = dev_get_drvdata(dev);
 
-	/* Ask ULPI to turn OFF Vbus */
-	dwc3_simple_vbus(simple->dwc, true);
+	if (!simple->wakeup_capable && !simple->dwc->is_d3) {
+		/* Ask ULPI to turn OFF Vbus */
+		dwc3_simple_vbus(simple->dwc, true);
+
+		/* Disable the clocks */
+		clk_bulk_disable(simple->num_clocks, simple->clks);
+	}
 
 	if (simple->need_reset)
 		reset_control_assert(simple->resets);
-- 
2.31.1

